pimd/pim6_mroute.c

```diff
@@ -0,0 +1,1120 @@
+			   pim_ifp->mroute_if_index, ifp->name,
+			   pim_ifp->pim->vrf->name);
+
+	memset(&vc, 0, sizeof(vc));
+	vc.mif6c_mifi = pim_ifp->mroute_if_index;
+	vc.mif6c_flags = flags;
+	vc.vifc_threshold = PIM_MROUTE_MIN_TTL;
+	vc.vifc_rate_limit = 0;
+
+	err = setsockopt(pim_ifp->pim->mroute_socket, IPPROTO_IPV6, MRT6_ADD_MIF,
+			 (void *)&vc, sizeof(vc));
+	if (err) {
+		zlog_warn(
+			"%s: failure: setsockopt(fd=%d,IPPROTO_IPV6,MRT6_ADD_MIF,mif_index=%d,ifaddr=%pPAs,flag=%d): errno=%d: %s",
+			__func__, pim_ifp->pim->mroute_socket, ifp->ifindex,
+			&ifaddr, flags, errno, safe_strerror(errno));
+		return -2;
+	}
+
+	return 0;
+}
+
+int pim_mroute_del_if(struct interface *ifp)
+{
+	struct pim_interface *pim_ifp = ifp->info;
+	struct mif6ctl vc;
+	int err;
+
+	if (PIM_DEBUG_MROUTE)
+		zlog_debug("%s: Del Mif %d (%s[%s])", __func__,
+			   pim_ifp->mroute_if_index, ifp->name,
+			   pim_ifp->pim->vrf->name);
+
+	memset(&vc, 0, sizeof(vc));
+	vc.mif6c_mifi = pim_ifp->mroute_if_index;
+
+	err = setsockopt(pim_ifp->pim->mroute_socket, IPPROTO_IPV6, MRT6_DEL_MIF,
+			 (void *)&vc, sizeof(vc));
+	if (err) {
+		zlog_warn(
+			"%s %s: failure: setsockopt(fd=%d,IPPROTO_IPV6,MRT6_DEL_MIF,mif_index=%d): errno=%d: %s",
+			__FILE__, __func__, pim_ifp->pim->mroute_socket,
+			pim_ifp->mroute_if_index, errno, safe_strerror(errno));
+		return -2;
+	}
+
+	return 0;
+}
+
+/*
+ * Prevent creating MFC entry with OIF=IIF.
+ *
+ * This is a protection against implementation mistakes.
+ *
+ * PIM protocol implicitely ensures loopfree multicast topology.
+ *
+ * IGMP must be protected against adding looped MFC entries created
+ * by both source and receiver attached to the same interface. See
+ * TODO T22.
+ * We shall allow igmp to create upstream when it is DR for the intf.
+ * Assume RP reachable via non DR.
+ */
+bool pim_mroute_allow_iif_in_oil(struct channel_oil *c_oil,
+		int oif_index)
+{
+#ifdef PIM_ENFORCE_LOOPFREE_MFC
+	struct interface *ifp_out;
+	struct pim_interface *pim_ifp;
+
+	if (c_oil->up &&
+		PIM_UPSTREAM_FLAG_TEST_ALLOW_IIF_IN_OIL(c_oil->up->flags))
+		return true;
+
+	ifp_out = pim_if_find_by_mcast_if_index(c_oil->pim, oif_index);
+	if (!ifp_out)
+		return false;
+	pim_ifp = ifp_out->info;
+	if (!pim_ifp)
+		return false;
+	if ((c_oil->oif_flags[oif_index] & PIM_OIF_FLAG_PROTO_IGMP) &&
+			PIM_I_am_DR(pim_ifp))
+		return true;
+
+	return false;
+#else
+	return true;
+#endif
+}
+
+static inline void pim_mroute_copy(struct mf6cctl *oil,
+		struct channel_oil *c_oil)
+{
+	int i;
+
+	oil->mf6cc_origin = c_oil->oil.mf6cc_origin;
+	oil->mf6cc_mcastgrp = c_oil->oil.mf6cc_mcastgrp;
+	oil->mf6cc_parent = c_oil->oil.mf6cc_parent;
+
+	for (i = 0; i < MAXMIFS; ++i) {
+		if ((oil->mf6cc_parent == i) &&
+				!pim_mroute_allow_iif_in_oil(c_oil, i)) {
+			IF_CLR(i, &oil->mf6cc_ifset);
+			continue;
+		}
+
+		if (c_oil->oif_flags[i] & PIM_OIF_FLAG_MUTE)
+			IF_CLR(i, &oil->mf6cc_ifset);
+		else
+			IF_SET(i, &oil->mf6cc_ifset); 
```

Comment:
This is going to set for all the OIFs. You need to add else if to check if c_oil->oif_flags[I] is set then set mf6cc_ifset